### ISSUE TITLE
Change in cookie handling with --as-ppx in 0.16 vs 0.15

### ISSUE THREAD

**Issue by jberdine** [ocaml-ppx/ppxlib#175]
I have been hunting down an issue that has turned up when upgrading to version 0.16, that seems to be a difference of behavior regarding how cookies are handles when `--as-ppx` is passed, e.g. by merlin. It doesn't seem like this is an intended consequence of the simplifications from moving to omp 2, but maybe.

The high level is that if the dune file contains something like `FLG -ppx '.../ppx.exe --as-ppx --cookie '\''library-name="mk_lib"'\'' ...'`, then merlin does not see the `library-name` cookie as set.

Prior to https://github.com/ocaml-ppx/ppxlib/pull/164, the `--cookie` flag would call `Driver.Cookies.set`, which would call `Migrate_parsetree.Driver.set_cookie` which stored the association in a hashtbl local to `Migrate_parsetree.Driver`. Now, `Driver.Cookies.set` calls `Ast_mapper.set_cookie`, which stores the association in a map local to `Ast_mapper`. The `get` functions are appropriately changed to.

This seems to work as expected normally, but not when `--as-ppx` is passed. In that case, `Driver.standalone_run_as_ppx_rewriter` is called, which calls `Ast_mapper.apply`, which calls `PpxContext.restore` and overwrites the map that stores the cookies. Note that previously, the cookies were not stored in the map that `PpxContext.restore` resets, but in the one that was local to `Migrate_parsetree.Driver`.

This isn't an intended change, right?

**Comment by jberdine** [comment]
@pitag-ha @jeremiedimino It seems that this may have been fixed by https://github.com/ocaml-ppx/ppxlib/pull/209 . Would that make sense to you, or just a lucky coincidence?

**Comment by jeremiedimino** [comment]
It make complete senses. I discovered the same issue through another report on a ppx rewriter. We discussed it and @pitag-ha added a test for it in #208: https://github.com/ocaml-ppx/ppxlib/pull/208/files#diff-e35b7f4f3668e1c5ad9c9633f2bc37cab051e831080cf0f30675f7d1263cb8d0. The test was then fixed by #209. We must have missed this ticket.

This code used to be proper spaghetti code. The execution started in ppxlib, then was going through `Ast_mapper` and then through the ocaml-migrate-parsetree driver and then back to ppxlib. We simplified all this and now we do everything in ppxlib, which as a side effect made it much easier to fix this issue.
